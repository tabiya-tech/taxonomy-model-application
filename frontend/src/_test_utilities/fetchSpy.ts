export function setupFetchSpy(
  expectedStatus: number,
  expectedResponseBody: string | object | undefined,
  contentType: "" | "application/json;charset=UTF-8"
): jest.SpyInstance {
  const responseBody =
    typeof expectedResponseBody === "string" ? expectedResponseBody : JSON.stringify(expectedResponseBody);
  const expectedResponse = new Response(responseBody, {
    status: expectedStatus,
    headers: { "Content-Type": contentType },
  });
  return jest.spyOn(window, "fetch").mockResolvedValue(expectedResponse);
}
