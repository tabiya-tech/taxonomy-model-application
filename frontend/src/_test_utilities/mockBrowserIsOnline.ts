import { fireEvent } from "@testing-library/react";

export function mockBrowserIsOnLine(isOnline: boolean) {
  if (window.navigator.onLine !== isOnline) {
    jest.spyOn(window.navigator, "onLine", "get").mockReturnValue(isOnline);
    if (isOnline) {
      fireEvent.online(window);
    } else {
      fireEvent.offline(window);
    }
  }
}

export function unmockBrowserIsOnLine() {
  jest.spyOn(window.navigator, "onLine", "get").mockRestore();
}
