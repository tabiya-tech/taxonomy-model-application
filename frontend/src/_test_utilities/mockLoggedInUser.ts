import { AuthContextValue, TabiyaUser, TabiyaUserRole, authContextDefaultValue } from "src/auth/AuthProvider";

export function mockLoggedInUser(mockedValues: Partial<AuthContextValue>){
  let user: TabiyaUser = mockedValues.user || { username: "foo", roles: [] };
  jest.spyOn(require("src/auth/hooks/useAuthUser"), "useAuthUser").mockReturnValue({
    ...authContextDefaultValue,
    user: user,
    setUser: (user: TabiyaUser) => null,
    hasRole: (role: TabiyaUserRole) => user.roles.includes(role),
    updateUserByAccessToken: () => {},
    ...mockedValues
  });
}
