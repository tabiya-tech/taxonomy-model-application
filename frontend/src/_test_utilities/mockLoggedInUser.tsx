import {
  AuthContextValue,
  TabiyaUser,
  authContextDefaultValue,
  AuthContext,
  AuthProviderProps,
} from "src/auth/AuthProvider";
import AuthAPISpecs from "api-specifications/auth";

export function mockLoggedInUser(mockedValues: Partial<AuthContextValue>) {
  let user: TabiyaUser = mockedValues.user || TestUsers.Anonymous;

  const value = {
    ...authContextDefaultValue,
    user: user,
    setUser: (user: TabiyaUser) => null,
    hasRole: (role: AuthAPISpecs.Enums.TabiyaRoles) => user.roles.includes(role),
    updateUserByAccessToken: () => {},
    ...mockedValues,
  };

  // @ts-ignore
  jest.spyOn(require("src/auth/AuthProvider"), "AuthProvider").mockImplementation((props: AuthProviderProps) => {
    return <AuthContext.Provider value={value}>{props.children}</AuthContext.Provider>;
  });
}

export const TestUsers = {
  ModelManager: {
    username: "ModelManagerUser",
    roles: [AuthAPISpecs.Enums.TabiyaRoles.MODEL_MANAGER],
  },
  RegisteredUser: {
    username: "RegisteredUser",
    roles: [AuthAPISpecs.Enums.TabiyaRoles.REGISTERED_USER],
  },
  Anonymous: {
    username: "AnonymousUser",
    roles: [AuthAPISpecs.Enums.TabiyaRoles.ANONYMOUS],
  },
};
