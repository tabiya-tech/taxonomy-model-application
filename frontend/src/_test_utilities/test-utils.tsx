// Based on https://testing-library.com/docs/react-testing-library/setup/

import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@mui/material";
import applicationTheme, { ThemeMode } from "src/theme/applicationTheme/applicationTheme";
import SnackbarProvider from "src/theme/SnackbarProvider/SnackbarProvider";
import { IsOnlineProvider } from "src/app/providers";

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <IsOnlineProvider>
      <ThemeProvider theme={applicationTheme(ThemeMode.LIGHT)}>
        <SnackbarProvider>{children}</SnackbarProvider>
      </ThemeProvider>
    </IsOnlineProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
