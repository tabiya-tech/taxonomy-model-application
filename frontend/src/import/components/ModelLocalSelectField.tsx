import { FormControl, Stack, MenuItem, Select, FormLabel, SelectChangeEvent, useTheme } from "@mui/material";
import React, { useEffect } from "react";
import Locale from "api-specifications/locale";
import { isUnspecified } from "src/utils/isUnspecified";

export const TEXT = {
  MODEL_LOCALE_SELECT_LABEL: "Select Model Locale",
};
const uniqueId = "dfdbbef3-ec63-44f4-9945-a757d9c81269";

export const DATA_TEST_ID = {
  MODEL_LOCALE_SELECT_FIELD: `model-locale-select-field-${uniqueId}`,
  MODEL_LOCALE_LABEL: `model-locale-label-${uniqueId}`,
  MODEL_LOCALE_INPUT: `model-locale-input-${uniqueId}`,
  MODEL_LOCALE_DROPDOWN: `model-locale-input-dropdown-${uniqueId}`,
  MODEL_LOCALE_ITEM: `model-locale-item-${uniqueId}`,
};

export interface ModelLocaleSelectProps {
  locales: Locale.Types.Payload[];
  notifyModelLocaleChanged?: (locale: Locale.Types.Payload) => any;
}

const ModelLocalSelectField = (props: Readonly<ModelLocaleSelectProps>) => {
  if (props.locales.length === 0 || props.locales.find((locale) => isUnspecified(locale.UUID))) {
    console.error("Locales should have at least one item");
  }

  const [selectedLocaleUUID, setSelectedLocaleUUID] = React.useState<string>(props.locales[0].UUID);

  const handleChange = (event: SelectChangeEvent) => {
    const UUID = event.target.value;
    setSelectedLocaleUUID(UUID);
    callNotifyModelLocaleChanged(UUID);
  };

  useEffect(() => {
    // notify the about the initial selected value, only once,
    // when the component is rendered for the first time
    callNotifyModelLocaleChanged(selectedLocaleUUID);
  });

  function callNotifyModelLocaleChanged(UUID: string) {
    if (props.notifyModelLocaleChanged) {
      const selected = props.locales.find((locale) => locale.UUID === UUID);
      if (selected) {
        props.notifyModelLocaleChanged(selected);
      }
    }
  }

  const theme = useTheme();
  return (
    <FormControl sx={{ width: "100%" }} data-testid={DATA_TEST_ID.MODEL_LOCALE_SELECT_FIELD}>
      <Stack spacing={theme.tabiyaSpacing.xs}>
        <FormLabel required id={uniqueId} data-testid={DATA_TEST_ID.MODEL_LOCALE_LABEL}>
          {TEXT.MODEL_LOCALE_SELECT_LABEL}
        </FormLabel>
        <Select
          labelId={uniqueId}
          data-testid={DATA_TEST_ID.MODEL_LOCALE_DROPDOWN}
          sx={{ width: "100%" }}
          variant={"standard"}
          value={selectedLocaleUUID}
          onChange={handleChange}
          inputProps={{
            "data-testid": DATA_TEST_ID.MODEL_LOCALE_INPUT,
          }}
        >
          {props.locales?.map((locale) => (
            <MenuItem data-testid={DATA_TEST_ID.MODEL_LOCALE_ITEM} key={locale.UUID} value={locale.UUID}>
              {locale.name} &nbsp; ({locale.shortCode})
            </MenuItem>
          ))}
        </Select>
      </Stack>
    </FormControl>
  );
};

export default ModelLocalSelectField;
