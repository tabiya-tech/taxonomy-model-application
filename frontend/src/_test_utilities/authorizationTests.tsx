import { render, screen } from "src/_test_utilities/test-utils";
import AuthAPISpecs from "api-specifications/auth";
import { mockLoggedInUser } from "./mockLoggedInUser";

export const ALL_USERS = [
  AuthAPISpecs.Enums.TabiyaRoles.ANONYMOUS,
  AuthAPISpecs.Enums.TabiyaRoles.REGISTERED_USER,
  AuthAPISpecs.Enums.TabiyaRoles.MODEL_MANAGER,
];

type AuthorizationTestsParams = {
  name: string;
  Component: JSX.Element;
  testIds?: string[];
  roles?: AuthAPISpecs.Enums.TabiyaRoles[];
  tests?: Partial<
    Record<
      AuthAPISpecs.Enums.TabiyaRoles,
      {
        visible?: string[];
        hidden?: string[];
        visibleDisabled?: string[];
      }
    >
  >;
};

export const authorizationTests = {
  defaultName: "Test the rendering of the component based on Roles",
  callback: ({ name, Component, testIds = [], roles = ALL_USERS, tests }: AuthorizationTestsParams) => {
    return () => {
      test.each(roles?.map((role) => [role]))(`${name} renders correctly for user with role %s`, (role) => {
        const commonElements = testIds || [];
        const { visible = [], visibleDisabled = [], hidden = [] } = tests?.[role] || {};

        mockLoggedInUser({ user: { username: "foo", roles: [role] } });

        render(Component);

        commonElements.forEach((testId) => expect(screen.getByTestId(testId)).toBeInTheDocument());

        visible?.forEach((testId) => {
          expect(screen.getByTestId(testId)).toBeVisible();
        });

        hidden?.forEach((testId) => {
          expect(screen.queryByTestId(testId)).not.toBeInTheDocument();
        });

        visibleDisabled?.forEach((testId) => {
          const element = screen.getByTestId(testId);
          expect(element).toBeVisible();
          if (element instanceof HTMLInputElement) {
            expect(element).toBeDisabled();
          }
        });
      });
    };
  },
};
