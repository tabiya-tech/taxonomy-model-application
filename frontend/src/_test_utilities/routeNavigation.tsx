import React from "react";
import { render, screen } from "src/_test_utilities/test-utils";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import userEvent from "@testing-library/user-event";

export function testNavigateToPath(
  givenComponentWithLink: React.ReactElement,
  givenElementName: string,
  givenElementToClickTestId: string,
  expectedPath: string
): void {
  return test(`clicking the '${givenElementName}' should navigate to '${expectedPath}'`, async () => {
    // GIVEN the givenComponentWithLink has access to a router
    // AND there is some initial route defined with some associated component
    // AND there component associated with the expectedPath
    const givenRouter = (
      <MemoryRouter initialEntries={["/foo-initial-path"]}>
        <>{givenComponentWithLink}</>
        <Routes>
          <Route path="/foo-initial-path" element={<div data-testid="/foo-initial-path"></div>} />
          <Route path={expectedPath} element={<div data-testid={expectedPath}></div>} />
        </Routes>
      </MemoryRouter>
    );
    // AND the givenRouter is rendered
    render(givenRouter);
    // AND the initial component is shown;
    expect(screen.getByTestId("/foo-initial-path")).toBeInTheDocument();

    // WHEN the element with the givenElementToClickTestId test idis clicked.
    const actualLink = screen.getByTestId(givenElementToClickTestId);
    await userEvent.click(actualLink);

    // THEN to component associated with the expectedPath is shown
    const actualComponent = screen.getByTestId(expectedPath);
    expect(actualComponent).toBeInTheDocument();
  });
}
