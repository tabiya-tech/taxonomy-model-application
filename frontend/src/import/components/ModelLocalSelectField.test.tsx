// mute the console
import 'src/_test_utilities/consoleMock';

import {render, screen, within} from "src/_test_utilities/test-utils";
import ModelLocalSelectField, {DATA_TEST_ID, TEXT} from "./ModelLocalSelectField";
import Locale from "api-specifications/locale";
import React from "react";
import userEvent from "@testing-library/user-event";

// Given a list of locales
const givenLocales: Locale.Types.Payload[] = [
  {
    UUID: "1",
    shortCode: "ZA",
    name: "South Africa"
  }, {
    UUID: "2",
    shortCode: "KA",
    name: "Kenya"
  }
];
describe("Locale Selector Component render tests", () => {
  beforeEach(() => {
    (console.error as jest.Mock).mockClear();
    (console.warn as jest.Mock).mockClear();
  });

  it("should render the locale selector component", () => {

    // WHEN the modelSelectionComponent is rendered
    render(<ModelLocalSelectField locales={givenLocales}/>)
    const modelLocalSelectField = screen.getByTestId(DATA_TEST_ID.MODEL_LOCALE_SELECT_FIELD);

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();
    // AND the local selector to be in the document
    expect(modelLocalSelectField).toBeInTheDocument();

    // AND expect the selected value to be the first from the given locales
    const dropdownElement = screen.getByTestId(DATA_TEST_ID.MODEL_LOCALE_DROPDOWN)
    const selectedInput = within(dropdownElement).getByTestId(DATA_TEST_ID.MODEL_LOCALE_INPUT);
    expect(selectedInput).toHaveAttribute("value", givenLocales[0].UUID);

    // AND expect label to be visible
    const label: HTMLLabelElement = screen.getByTestId(DATA_TEST_ID.MODEL_LOCALE_LABEL)
    expect(label).toBeInTheDocument();
    expect(label).toBeVisible();
    // AND expect label to have correct text
    expect(label).toHaveTextContent(TEXT.MODEL_LOCALE_SELECT_LABEL)
  })

  test("multiple components should have a unique id", () => {
    // GIVEN multiple selectFieldComponents are selected
    render(
      <div>
        <ModelLocalSelectField locales={givenLocales}/>
        <ModelLocalSelectField locales={givenLocales}/>
      </div>
    );

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();
    // AND expect each dropdownElement to have different ids
    const labelElements = screen.getAllByTestId(DATA_TEST_ID.MODEL_LOCALE_LABEL);
    expect(labelElements.length).toBe(2);
    expect(labelElements[0].id).not.toBe(labelElements[1].id);
  });
});

describe("Locale Selector Component actions tests only", () => {
  it("should call the notifyChangeHandler when default selected value", async () => {
    // GIVEN a notifyChangeHandler
    const givenNotifyChangeHandler = jest.fn();

    // AND the modelSelectionComponent is rendered with some model locales
    render(<ModelLocalSelectField locales={givenLocales} notifyModelLocaleChanged={givenNotifyChangeHandler}/>);

    // THEN expect the notifyChangeHandler to be called with the selected locale
    expect(givenNotifyChangeHandler).toHaveBeenLastCalledWith(givenLocales[0]);
  })

  it("should call the notifyChangeHandler when a locale is selected", async () => {
    // GIVEN a notifyChangeHandler
    const givenNotifyChangeHandler = jest.fn();

    // AND the modelSelectionComponent is rendered with some model locales
    render(<ModelLocalSelectField locales={givenLocales} notifyModelLocaleChanged={givenNotifyChangeHandler}/>);

    const dropdownElement = screen.getByTestId(DATA_TEST_ID.MODEL_LOCALE_DROPDOWN)
    const button = within(dropdownElement).getByRole('button');
    await userEvent.click(button);

    // THEN expect the dropdown to be open
    const dropdownList = screen.getAllByTestId(DATA_TEST_ID.MODEL_LOCALE_ITEM);
    expect(dropdownList).toHaveLength(2);

    const secondLocaleElement = dropdownList.find((item) => item.getAttribute('data-value') === givenLocales[1].UUID);
    if(secondLocaleElement){
      await userEvent.click(secondLocaleElement);
    }
    // THEN expect the notifyChangeHandler to be called with the selected locale
    expect(givenNotifyChangeHandler).toHaveBeenLastCalledWith(givenLocales[1]);
  })
})