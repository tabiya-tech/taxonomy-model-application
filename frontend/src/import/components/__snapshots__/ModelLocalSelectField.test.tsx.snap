// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Locale Selector Component render tests should render the locale selector component 1`] = `
<div
  class="MuiFormControl-root css-q8hpuo-MuiFormControl-root"
  data-testid="model-locale-select-field-dfdbbef3-ec63-44f4-9945-a757d9c81269"
>
  <div
    class="MuiStack-root css-ri2h3u-MuiStack-root"
  >
    <label
      class="MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required css-1k0yz6l-MuiFormLabel-root"
      data-testid="model-locale-label-dfdbbef3-ec63-44f4-9945-a757d9c81269"
      id="dfdbbef3-ec63-44f4-9945-a757d9c81269"
    >
      Select Model Locale
      <span
        aria-hidden="true"
        class="MuiFormLabel-asterisk css-1th3aq5-MuiFormLabel-asterisk"
      >
         
        *
      </span>
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl  css-3sxds7-MuiInputBase-root-MuiInput-root-MuiSelect-root"
      data-testid="model-locale-input-dropdown-dfdbbef3-ec63-44f4-9945-a757d9c81269"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="dfdbbef3-ec63-44f4-9945-a757d9c81269"
        class="MuiSelect-select MuiSelect-standard MuiInputBase-input MuiInput-input css-wrly8k-MuiSelect-select-MuiInputBase-input-MuiInput-input"
        role="button"
        tabindex="0"
      >
        South Africa
           (
        ZA
        )
      </div>
      <input
        aria-hidden="true"
        aria-invalid="false"
        class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        data-testid="model-locale-input-dfdbbef3-ec63-44f4-9945-a757d9c81269"
        tabindex="-1"
        value="1"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconStandard css-b9pvgf-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
</div>
`;
